### PR TITLE
HDDS-7826. Add support for application to probe output stream capability

### PR DESCRIPTION
**Pull request by jojochuang**
## What changes were proposed in this pull request?

Add StreamCapability interface to Ozone outputstream.

## What is the link to the Apache JIRA

https://issues.apache.org/jira/browse/HDDS-7826

## How was this patch tested?

Unit tests

### CODE REVIEW

**Comment by jojochuang** [comment]
Posting the PR as a draft, because it depends on the completion of hsync API. A contract test will actually fail because hsync doesn't make the file visible.

**Comment by kerneltime** [comment]
@aswinshakil can you please take a look?

File: hadoop-ozone/ozonefs-common/src/main/java/org/apache/hadoop/fs/ozone/CapableOzoneFSOutputStream.java [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+      case OzoneStreamCapabilities.HSYNC:
+        // TODO: switch to true after HDDS-7688
+        return false;
+      default:
+        return false;
+      }
+    }
+    // deal with CryptoOutputStream
+    return StoreImplementationUtils.hasCapability(
+      getWrappedOutputStream().getOutputStream(), capability);
```
Comment (szetszwo) [inline hadoop-ozone/ozonefs-common/src/main/java/org/apache/hadoop/fs/ozone/CapableOzoneFSOutputStream.java:61]: We may use `os` here.

**Review by szetszwo — APPROVED** [review]
+1 the change looks good.

File: hadoop-ozone/ozonefs-common/src/main/java/org/apache/hadoop/fs/ozone/OzoneStreamCapabilities.java [hunk 2]
```diff
@@ -44,4 +44,8 @@ private OzoneStreamCapabilities() {
    * TODO: If Hadoop dependency is upgraded, this string can be removed.
    */
   static final String UNBUFFER = "in:unbuffer";
+
+  static final String HSYNC = "hsync";
+
+  static final String HFLUSH = "hflush";
```
Comment (adoroszlai) [inline hadoop-ozone/ozonefs-common/src/main/java/org/apache/hadoop/fs/ozone/OzoneStreamCapabilities.java:50]: Only if you update the patch anyway:

I think we can use these constants from Hadoop's `StreamCapabilities` directly.  `OzoneStreamCapabilities` was added when Ozone used Hadoop 3.2, which didn't have the other two constants.  (This reminds me to address the TODO above, filed HDDS-7863 for that.)

**Comment by adoroszlai** [comment]
Thanks @jojochuang for updating the patch, @szetszwo for the review.

**Pull request merged**